### PR TITLE
Replaced assert statement with if statement and a meaningfull error message

### LINKED ISSUE #150: Warn of failures in aligned_alloc
https://github.com/microsoft/DiskANN/blob/8635dadb63391c09da0e450278136de5d88ad7cc/include/utils.h#L177

Replace assert with if and a meaningful message.

### PR DESCRIPTION
**Pull request by ChakPC**
- [ ] Does this PR have a descriptive title that could go in our release notes?
- [ ] Does this PR add any new dependencies?
- [ ] Does this PR modify any existing APIs?
   - [ ] Is the change to the API backwards compatible?
- [ ] Should this result in any changes to our documentation, either updating existing docs or adding new ones?
 
#### Reference Issues/PRs
Fixes #150. 

#### What does this implement/fix? Briefly explain your changes.
This PR fixes #150 by replacing assert statement by an if statement with a meaningful error message about memory allocation failure. Further, an exception is also thrown so that the program terminates in a similar fashion as in the case of assert statement.

### CODE REVIEW

File: include/utils.h [hunk 1]
```diff
@@ -160,7 +160,13 @@ namespace diskann {
 #else
     *ptr = ::_aligned_malloc(size, align);  // note the swapped arguments!
 #endif
-    assert(*ptr != nullptr);
+    if (*ptr == nullptr) {
+      std::stringstream stream;
+      stream << "Memory Allocation Failed.";
+      diskann::cerr << stream.str()
+                    << std::endl;
```
Comment (harsha-simhadri) [inline include/utils.h:167]: Could you please run clang format.

File: include/utils.h [hunk 2]
```diff
@@ -160,7 +160,13 @@ namespace diskann {
...
+      throw diskann::ANNException(stream.str(), -1);
```
Comment (harsha-simhadri) [inline include/utils.h:168]: add __FUNCSIG__, __FILE__, __LINE__ as additional parameters to the exception

**Review by harsha-simhadri — CHANGES_REQUESTED** [review]
Does this fix issue 150 at all allocation failure points or just this one?

File: include/utils.h (hunk 1 above)
Comment (ChakPC) [reply include/utils.h]: Done in the new commit.

File: include/utils.h (hunk 2 above)
Comment (ChakPC) [reply include/utils.h]: Added in new commit.

**Comment by ChakPC** [comment]
> Does this fix issue 150 at all allocation failure points or just this one?

In the new commit, it resolves all memory allocation failure points in include/utils.h file. Moreover, there were some other assert statements which checked alignment of requested memory size for allocation. Those assert statements are also converted to if statements. Instead of repeating the error printing message in each if condition, the concerned code snippet is moved to a function and that function is invoked in the if statement.

**Comment by ChakPC** [comment]
> @ChakPC please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

File: include/utils.h [hunk 3]
```diff
@@ -152,27 +151,49 @@ class AlignedFileReader;
...
+                                __LINE__);
+  }
+
+  inline void report_memory_allocation_failure() {
+    std::stringstream stream;
+    stream << "Memory Allocation Failed.";
+    print_error_and_terminate(stream);
+  }
+
+  inline void report_misallignment_of_requested_size() {
```
Comment (harsha-simhadri) [inline include/utils.h:166]: typo.. misalignment instead of misallignment (single l)

File: include/utils.h [hunk 4]
```diff
@@ -152,27 +151,49 @@ class AlignedFileReader;
...
+    std::stringstream stream;
+    stream << "Requested memory size is misalligned and can not be allocated.";
```
Comment (harsha-simhadri) [inline include/utils.h:168]: correct typo in misalligned.
Add info on "Requsted memory size is not a multiple of {align}. Can not be allocated". Print value of align so it's easier to interpret error.

**Review by harsha-simhadri — CHANGES_REQUESTED** [review]

File: include/utils.h (hunk 3 above)
Comment (ChakPC) [reply include/utils.h]: Corrected.

File: include/utils.h (hunk 4 above)
Comment (ChakPC) [reply include/utils.h]: Added the info about align in new commit.

**Review by harsha-simhadri — APPROVED** [review]

**Pull request merged**